thingsboard_gateway/extensions/tcpserver/tcpserver_connector.py

```diff
@@ -0,0 +1,243 @@
+                                   "deviceType": converted_data["deviceType"]}
+                        if to_send.get("telemetry") is None:
+                            to_send["telemetry"] = []
+                        if to_send.get("attributes") is None:
+                            to_send["attributes"] = []
+                        for telemetry_dict in converted_data["timeseries"]:
+                            for key, value in telemetry_dict.items():
+                                if device_config["last_telemetry"].get(key) is None or \
+                                        device_config["last_telemetry"][key] != value:
+                                    device_config["last_telemetry"][key] = value
+                                    to_send["telemetry"].append({key: value})
+                        for attribute_dict in converted_data["attributes"]:
+                            for key, value in attribute_dict.items():
+                                if device_config["last_attributes"].get(key) is None or \
+                                        device_config["last_attributes"][key] != value:
+                                    device_config["last_attributes"][key] = value
+                                    to_send["attributes"].append({key: value})
+                        if to_send.get("attributes") or to_send.get("telemetry"):
+                            TCPServer_gateway.send_to_storage(self.get_name(), to_send)
+                        else:
+                            log.debug("Data has not been changed.")
+                    elif converted_data and device_config["config"].get(
+                            "sendDataOnlyOnChange") is None or not device_config["config"].get(
+                            "sendDataOnlyOnChange"):
+                        to_send = {"deviceName": converted_data["deviceName"],
+                                   "deviceType": converted_data["deviceType"]}
+                        # if converted_data["telemetry"] != self.__devices[device]["telemetry"]:
+                        device_config["last_telemetry"] = converted_data["timeseries"]
+                        to_send["telemetry"] = converted_data["timeseries"]
+                        # if converted_data["attributes"] != self.__devices[device]["attributes"]:
+                        device_config["last_attributes"] = converted_data["attributes"]
+                        to_send["attributes"] = converted_data["attributes"]
+                        TCPServer_gateway.send_to_storage(self.get_name(), to_send)
+
+class TCPServerConnector(Connector, threading.Thread):
+    def __init__(self, gateway, config, connector_type):
+        super().__init__()
+        self.statistics = {'MessagesReceived': 0,
+                           'MessagesSent': 0}
+        self.__gateway = gateway
+        self._connector_type = connector_type
+        self.__server = None
+        self.__config = config
+
+        global TCPServer_gateway
```

Comment:
Please don't use the global variables, it can cause some issues with remote configuration.

Reply:
I ma sorry,  I don’t know how to pass the parameters without using global variables.

Reply:
Well, then you can leave it like this, I will investigate this and refactor. I think we can move the declaration of the Handler class inside the connector class and have access to the variables of the connector object. But I'm not sure, and I need to do some research.

Reply:
ok. Thank you very much.